# PREPROCESSING OF THE DATA

### Data is cleaned and organized for building and training the model

# IMPORTING NECESSARY LIBRARIES


In [1]:
import pandas as pd
import numpy as np
import os
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import style

np.random.seed(42)
from datetime import datetime, timedelta
%matplotlib inline

ModuleNotFoundError: No module named 'seaborn'

# READING THE DATASETS

In [5]:
customers_db = pd.read_csv("C:\OLIST\olist_customers_dataset.csv")
order_items_db = pd.read_csv("C:\OLIST\olist_order_items_dataset.csv")
order_payments_db = pd.read_csv("C:\OLIST\olist_order_payments_dataset.csv")
orders_db = pd.read_csv("C:\OLIST\olist_orders_dataset.csv")
products_db = pd.read_csv("C:\OLIST\olist_products_dataset.csv")
sellers_db = pd.read_csv("C:\OLIST\olist_sellers_dataset.csv")
category_db = pd.read_csv("C:\OLIST\product_category_name_translation.csv")
geolocation_db = pd.read_csv("C:\OLIST\olist_geolocation_dataset.csv")
product_reviews_db = pd.read_csv("C:\OLIST\olist_order_reviews_dataset.csv")

#customers_db = pd.read_csv("/Users/tejaltandel/archive/olist_customers_dataset.csv")
#geolocation_db = pd.read_csv("/Users/tejaltandel/archive/olist_geolocation_dataset.csv")
#order_items_db = pd.read_csv("/Users/tejaltandel/archive/olist_order_items_dataset.csv")
#order_payments_db = pd.read_csv("/Users/tejaltandel/archive/olist_order_payments_dataset.csv")
#product_reviews_db = pd.read_csv("/Users/tejaltandel/archive/olist_order_reviews_dataset.csv")
#orders_db = pd.read_csv("/Users/tejaltandel/archive/olist_orders_dataset.csv")
#products_db = pd.read_csv("/Users/tejaltandel/archive/olist_products_dataset.csv")
#sellers_db = pd.read_csv("/Users/tejaltandel/archive/olist_sellers_dataset.csv")
#category_db = pd.read_csv("/Users/tejaltandel/archive/product_category_name_translation.csv")

# REMOVAL OF DUMMY VALUES IN CUSTOMER DATASET

In [6]:
customers_db

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [7]:
nums = np.random.choice(range(1,99442), size = customers_db['customer_id'].nunique(), replace=False)

In [8]:
customers_db['customer_id2'] = customers_db['customer_id'].map(dict(zip(customers_db['customer_id'].unique(), nums)))

In [9]:
customers_db['customer_unique_id'].nunique()

96096

In [10]:
nums6 = np.random.choice(range(1,97000), size = customers_db['customer_unique_id'].nunique(), replace=False)

In [11]:
customers_db['uniqueid_2'] = customers_db['customer_unique_id'].map(dict(zip(customers_db['customer_unique_id'].unique(), nums6)))

In [12]:
customers_db.uniqueid_2 = customers_db.uniqueid_2.astype(str).str.pad(5,fillchar='0')

In [13]:
customers_db.uniqueid_2=['C'+x for x in customers_db.uniqueid_2 ]

In [14]:
# Get the DataFrame column names as a list
clist = list(customers_db.columns)

# Rearranging the list 
clist_new = clist[-1:]+clist[:-1]   # brings the last column in the first place

# Pass the new list to the DataFrame 
customers_db = customers_db[clist_new]
customers_db

uniqueid_2                       customer_id  \
0         C59074  06b8999e2fba1a1fbc88172c00ba8bc7   
1         C86482  18955e83d337fd6b2def6b18a428ac77   
2         C22074  4e7b3e00288586ebd08712fdd0374a03   
3         C54407  b2b6027bc5c5109e529d4dc6358b12c3   
4         C40086  4f2d8ab171c80ec8364f7c12e35b23ad   
...          ...                               ...   
99436     C79150  17ddf5dd5d51696bb3d7c6291687be6f   
99437     C43345  e7b71a9017aa05c9a7fd292d714858e8   
99438     C59186  5e28dfe12db7fb50a4b2f691faecea5e   
99439     C75844  56b18e2166679b8a959d72dd06da27f9   
99440     C06102  274fa6071e5e17fe303b9748641082c8   

                     customer_unique_id  customer_zip_code_prefix  \
0      861eff4711a542e4b93843c6dd7febb0                     14409   
1      290c77bc529b7ac935b93aa66c333dc3                      9790   
2      060e732b5b29e8181a18229c7b0b2b5e                      1151   
3      259dac757896d24d7702b9acbbff3f3c                      8775   
4      345ecd01c38d18a9036ed96c73b8d066                     13056   
...                                 ...                       ...   
99436  1a29b476fee25c95fbafc67c5ac95cf8                      3937   
99437  d52a67c98be1cf6a5c84435bd38d095d                      6764   
99438  e9f50caf99f032f0bf3c55141f019d99                     60115   
99439  73c2643a0a458b49f58cea58833b192e                     92120   
99440  84732c5050c01db9b23e19ba39899398                      6703   

               customer_city customer_state  customer_id2  
0                     franca             SP         96249  
1      sao bernardo do campo             SP         54293  
2                  sao paulo             SP         92207  
3            mogi das cruzes             SP         58987  
4                   campinas             SP            67  
...                      ...            ...           ...  
99436              sao paulo             SP         78724  
99437        taboao da serra             SP         16018  
99438              fortaleza             CE         91364  
99439                 canoas             RS         77476  
99440                  cotia             SP         93204  

[99441 rows x 7 columns]

In [21]:
customers_db.to_csv('C:\OLIST\olist_customers_dataset_new.csv',index=False)
#customers_db.to_csv('/Users/tejaltandel/archive/olist_customers_dataset.csv',index=False)

In [22]:
customers_db['customer_id'].nunique()

99441

In [23]:
boolean = customers_db['customer_id'].duplicated().any()

In [24]:
boolean

False

### Thus, unique record of customers is created and saved. Now we will be following the same procedure for all the other datasets

In [25]:
products_db


product_width_cm                        product_id  \
0                  14.0  1e9e8ef04dbcff4541ed26657ea517e5   
1                  20.0  3aa071139cb16b67ca9e5dea641aaa2f   
2                  15.0  96bd76ec8810374ed1b65e291975717f   
3                  26.0  cef67bcfe19066a932b7673e239eb23d   
4                  13.0  9dc1a7de274444849c219cff195d0b71   
...                 ...                               ...   
32946              40.0  a0b7d5a992ccda646f2d34e418fff5a0   
32947              16.0  bf4538d88321d0fd4412a93c974510e6   
32948              27.0  9a7c6041fa9592d9d9ef6cfe62a71f8c   
32949              20.0  83808703fc0706a22e264b9d75f04a2e   
32950               7.0  106392145fca363410d287a815be6de4   

                   product_category_name  product_name_lenght  \
0                             perfumaria                 40.0   
1                                  artes                 44.0   
2                          esporte_lazer                 46.0   
3                                  bebes                 27.0   
4                  utilidades_domesticas                 37.0   
...                                  ...                  ...   
32946                   moveis_decoracao                 45.0   
32947  construcao_ferramentas_iluminacao                 41.0   
32948                    cama_mesa_banho                 50.0   
32949             informatica_acessorios                 60.0   
32950                    cama_mesa_banho                 58.0   

       product_description_lenght  product_photos_qty  product_weight_g  \
0                           287.0                 1.0             225.0   
1                           276.0                 1.0            1000.0   
2                           250.0                 1.0             154.0   
3                           261.0                 1.0             371.0   
4                           402.0                 4.0             625.0   
...                           ...                 ...               ...   
32946                        67.0                 2.0           12300.0   
32947                       971.0                 1.0            1700.0   
32948                       799.0                 1.0            1400.0   
32949                       156.0                 2.0             700.0   
32950                       309.0                 1.0            2083.0   

       product_length_cm  product_height_cm  
0                   16.0               10.0  
1                   30.0               18.0  
2                   18.0                9.0  
3                   26.0                4.0  
4                   20.0               17.0  
...                  ...                ...  
32946               40.0               40.0  
32947               16.0               19.0  
32948               27.0                7.0  
32949               31.0               13.0  
32950               12.0                2.0  

[32951 rows x 9 columns]

In [26]:
nums2 = np.random.choice(range(1,32952), size = products_db['product_id'].nunique(), replace=False)

In [27]:
products_db['productid_2'] = products_db['product_id'].map(dict(zip(products_db['product_id'].unique(), nums2)))

In [28]:
products_db.productid_2 = products_db.productid_2.astype(str).str.pad(5,fillchar='0')

In [29]:
products_db.to_csv('C:\OLIST\olist_products_dataset_new.csv',index=False)
#products_db.to_csv('/Users/tejaltandel/archive/olist_products_dataset.csv',index=False)

In [30]:
products_db.productid_2=['P'+x for x in products_db.productid_2 ]

In [31]:
# Get the DataFrame column names as a list
plist = list(products_db.columns)

# Rearrange list the way you like 
plist_new = plist[-1:]+plist[:-1]   # brings the last column in the first place

# Pass the new list to the DataFrame 
products_db = products_db[plist_new]
products_db

productid_2  product_width_cm                        product_id  \
0          P09648              14.0  1e9e8ef04dbcff4541ed26657ea517e5   
1          P03118              20.0  3aa071139cb16b67ca9e5dea641aaa2f   
2          P02353              15.0  96bd76ec8810374ed1b65e291975717f   
3          P12294              26.0  cef67bcfe19066a932b7673e239eb23d   
4          P15781              13.0  9dc1a7de274444849c219cff195d0b71   
...           ...               ...                               ...   
32946      P11548              40.0  a0b7d5a992ccda646f2d34e418fff5a0   
32947      P30998              16.0  bf4538d88321d0fd4412a93c974510e6   
32948      P22009              27.0  9a7c6041fa9592d9d9ef6cfe62a71f8c   
32949      P27798              20.0  83808703fc0706a22e264b9d75f04a2e   
32950      P25149               7.0  106392145fca363410d287a815be6de4   

                   product_category_name  product_name_lenght  \
0                             perfumaria                 40.0   
1                                  artes                 44.0   
2                          esporte_lazer                 46.0   
3                                  bebes                 27.0   
4                  utilidades_domesticas                 37.0   
...                                  ...                  ...   
32946                   moveis_decoracao                 45.0   
32947  construcao_ferramentas_iluminacao                 41.0   
32948                    cama_mesa_banho                 50.0   
32949             informatica_acessorios                 60.0   
32950                    cama_mesa_banho                 58.0   

       product_description_lenght  product_photos_qty  product_weight_g  \
0                           287.0                 1.0             225.0   
1                           276.0                 1.0            1000.0   
2                           250.0                 1.0             154.0   
3                           261.0                 1.0             371.0   
4                           402.0                 4.0             625.0   
...                           ...                 ...               ...   
32946                        67.0                 2.0           12300.0   
32947                       971.0                 1.0            1700.0   
32948                       799.0                 1.0            1400.0   
32949                       156.0                 2.0             700.0   
32950                       309.0                 1.0            2083.0   

       product_length_cm  product_height_cm  
0                   16.0               10.0  
1                   30.0               18.0  
2                   18.0                9.0  
3                   26.0                4.0  
4                   20.0               17.0  
...                  ...                ...  
32946               40.0               40.0  
32947               16.0               19.0  
32948               27.0                7.0  
32949               31.0               13.0  
32950               12.0                2.0  

[32951 rows x 10 columns]

In [32]:
sellers_db


seller_id  seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                   13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4     51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...                                ...                     ...   
3090  98dddbc4601dd4443ca174359b237166                   87111   
3091  f8201cab383e484733266d1906e2fdfa                   88137   
3092  74871d19219c7d518d0090283e03c137                    4650   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

In [33]:
nums3 = np.random.choice(range(1,3096), size = sellers_db['seller_id'].nunique(), replace=False)

In [34]:
sellers_db['sellerid_2'] = sellers_db['seller_id'].map(dict(zip(sellers_db['seller_id'].unique(), nums3)))

In [35]:
sellers_db.sellerid_2 = sellers_db.sellerid_2.astype(str).str.pad(4,fillchar='0')

In [36]:
sellers_db.sellerid_2=['S'+x for x in sellers_db.sellerid_2 ]

In [37]:
# Get the DataFrame column names as a list
slist = list(sellers_db.columns)

# Rearrange list the way you like 
slist_new = slist[-1:]+slist[:-1]   # brings the last column in the first place

# Pass the new list to the DataFrame  
sellers_db = sellers_db[slist_new]
sellers_db

sellerid_2                         seller_id  seller_zip_code_prefix  \
0         S1699  3442f8959a84dea7ee197c632cb2df15                   13023   
1         S2188  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2         S2117  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3         S2455  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4         S2109  51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...         ...                               ...                     ...   
3090      S2782  98dddbc4601dd4443ca174359b237166                   87111   
3091      S0878  f8201cab383e484733266d1906e2fdfa                   88137   
3092      S0504  74871d19219c7d518d0090283e03c137                    4650   
3093      S1232  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094      S3006  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 5 columns]

In [39]:
sellers_db.to_csv('C:\OLIST\olist_sellers_dataset_new.csv',index=False)
#sellers_db.to_csv('/Users/tejaltandel/archive/olist_sellers_dataset.csv',index=False)

In [40]:
category_db

product_category_name product_category_name_english
0                    beleza_saude                 health_beauty
1          informatica_acessorios         computers_accessories
2                      automotivo                          auto
3                 cama_mesa_banho                bed_bath_table
4                moveis_decoracao               furniture_decor
..                            ...                           ...
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

[71 rows x 2 columns]

In [41]:
geolocation_db

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                               1037       -23.545621       -46.639292   
1                               1046       -23.546081       -46.644820   
2                               1046       -23.546129       -46.642951   
3                               1041       -23.544392       -46.639499   
4                               1035       -23.541578       -46.641607   
...                              ...              ...              ...   
1000158                        99950       -28.068639       -52.010705   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000161                        99980       -28.388932       -51.846871   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS  

[1000163 rows x 5 columns]

In [42]:
product_reviews_db

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [43]:
product_reviews_db["review_id"].nunique()

98410

In [44]:
boolean = product_reviews_db['review_id'].duplicated().any()

In [45]:
boolean

True

In [46]:
nums4 = np.random.choice(range(1,99999), size = product_reviews_db['review_id'].nunique(), replace=False)

In [47]:
product_reviews_db['review_id2'] = product_reviews_db['review_id'].map(dict(zip(product_reviews_db['review_id'].unique(), nums4)))

In [48]:
product_reviews_db.review_id2 = product_reviews_db.review_id2.astype(str).str.pad(5,fillchar='0')

In [49]:
product_reviews_db.review_id2=['PR'+x for x in product_reviews_db.review_id2 ]

In [50]:
# Get the DataFrame column names as a list
prlist = list(product_reviews_db.columns)

# Rearrange list the way you like 
prlist_new = prlist[-1:]+prlist[:-1]   # brings the last column in the first place

# Pass the new list to the DataFrame
product_reviews_db = product_reviews_db[prlist_new]
product_reviews_db

review_id2                         review_id  \
0        PR11290  7bc2406110b926393aa56f80a40eba40   
1        PR33714  80e641a11e56f04c1ad469d5645fdfde   
2        PR22105  228ce5500dc1d8e020d8d1322874b6f0   
3        PR90299  e64fb393e7b32834bb789ff8bb30750e   
4        PR01227  f7c4243c7fe1938f181bec41a392bdeb   
...          ...                               ...   
99219    PR15885  574ed12dd733e5fa530cfd4bbf39d7c9   
99220    PR27187  f3897127253a9592a73be9bdfdf4ed7a   
99221    PR39200  b3de70c89b1510c4cd3d0649fd302472   
99222    PR06310  1adeb9d84d72fe4e337617733eb85149   
99223    PR08109  efe49f1d6f951dd88b51e6ccd4cc548f   

                               order_id  review_score review_comment_title  \
0      73fc7af87114b39712e6da79b0a377eb             4                  NaN   
1      a548910a1c6147796b98fdf73dbeba33             5                  NaN   
2      f9e4b658b201a9f2ecdecbb34bed034b             5                  NaN   
3      658677c97b385a9be170737859d3511b             5                  NaN   
4      8e6bfb81e283fa7e4f11123a3fb894f1             5                  NaN   
...                                 ...           ...                  ...   
99219  2a8c23fee101d4d5662fa670396eb8da             5                  NaN   
99220  22ec9f0669f784db00fa86d035cf8602             5                  NaN   
99221  55d4004744368f5571d1f590031933e4             5                  NaN   
99222  7725825d039fc1f0ceb7635e3f7d9206             4                  NaN   
99223  90531360ecb1eec2a1fbb265a0db0508             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 8 columns]

In [52]:
product_reviews_db.to_csv('C:\OLIST\olist_order_reviews_dataset_new.csv',index=False)
#product_reviews_db.to_csv('/Users/tejaltandel/archive/olist_order_reviews_dataset.csv',index=False)

In [53]:
orders_db

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

In [54]:
orders_db['order_id'].nunique()

99441

In [55]:
nums5 = np.random.choice(range(1,99999), size = orders_db['order_id'].nunique(), replace=False)

In [56]:
orders_db['order_id2'] = orders_db['order_id'].map(dict(zip(orders_db['order_id'].unique(), nums5)))

In [57]:
orders_db.order_id2 = orders_db.order_id2.astype(str).str.pad(5,fillchar='0')

In [58]:
orders_db.order_id2=['ON'+x for x in orders_db.order_id2 ]

In [59]:
# Get the DataFrame column names as a list
olist = list(orders_db.columns)

# Rearrange list the way you like 
olist_new = olist[-1:]+olist[:-1]   # brings the last column in the first place

# Pass the new list to the DataFrame 
orders_db = orders_db[olist_new]
orders_db

order_id2                          order_id  \
0       ON80190  e481f51cbdc54678b7cc49136f2d6af7   
1       ON11081  53cdb2fc8bc7dce0b6741e2150273451   
2       ON05597  47770eb9100c2d0c44946d9cf07ec65d   
3       ON36385  949d5b44dbf5de918fe9c16f97b45f8a   
4       ON65566  ad21c59c0840e6cb83a9ceb5573f8159   
...         ...                               ...   
99436   ON65212  9c5dedf39a927c1b2549525ed64a053c   
99437   ON25205  63943bddc261676b46f01ca7ac2f7bd8   
99438   ON23669  83c1379a015df1e13d02aae0204711ab   
99439   ON25474  11c177c8e97725db2631073c19f07b62   
99440   ON02926  66dea50a8b16d9b4dee7af250b4be1a5   

                            customer_id order_status order_purchase_timestamp  \
0      9ef432eb6251297304e76186b10a928d    delivered      2017-10-02 10:56:33   
1      b0830fb4747a6c6d20dea0b8c802d7ef    delivered      2018-07-24 20:41:37   
2      41ce2a54c0b03bf3443c3d931a367089    delivered      2018-08-08 08:38:49   
3      f88197465ea7920adcdbec7375364d82    delivered      2017-11-18 19:28:06   
4      8ab97904e6daea8866dbdbc4fb7aad2c    delivered      2018-02-13 21:18:39   
...                                 ...          ...                      ...   
99436  39bd1228ee8140590ac3aca26f2dfe00    delivered      2017-03-09 09:54:05   
99437  1fca14ff2861355f6e5f14306ff977a7    delivered      2018-02-06 12:58:58   
99438  1aa71eb042121263aafbe80c1b562c9c    delivered      2017-08-27 14:46:43   
99439  b331b74b18dc79bcdf6532d51e1637c1    delivered      2018-01-08 21:28:27   
99440  edb027a75a1449115f6b43211ae02a24    delivered      2018-03-08 20:57:30   

         order_approved_at order_delivered_carrier_date  \
0      2017-10-02 11:07:15          2017-10-04 19:55:00   
1      2018-07-26 03:24:27          2018-07-26 14:31:00   
2      2018-08-08 08:55:23          2018-08-08 13:50:00   
3      2017-11-18 19:45:59          2017-11-22 13:39:59   
4      2018-02-13 22:20:29          2018-02-14 19:46:34   
...                    ...                          ...   
99436  2017-03-09 09:54:05          2017-03-10 11:18:03   
99437  2018-02-06 13:10:37          2018-02-07 23:22:42   
99438  2017-08-27 15:04:16          2017-08-28 20:52:26   
99439  2018-01-08 21:36:21          2018-01-12 15:35:03   
99440  2018-03-09 11:20:28          2018-03-09 22:11:59   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-10-10 21:25:13           2017-10-18 00:00:00  
1               2018-08-07 15:27:45           2018-08-13 00:00:00  
2               2018-08-17 18:06:29           2018-09-04 00:00:00  
3               2017-12-02 00:28:42           2017-12-15 00:00:00  
4               2018-02-16 18:17:02           2018-02-26 00:00:00  
...                             ...                           ...  
99436           2017-03-17 15:08:01           2017-03-28 00:00:00  
99437           2018-02-28 17:37:56           2018-03-02 00:00:00  
99438           2017-09-21 11:24:17           2017-09-27 00:00:00  
99439           2018-01-25 23:32:54           2018-02-15 00:00:00  
99440           2018-03-16 13:08:30           2018-04-03 00:00:00  

[99441 rows x 9 columns]

In [61]:
orders_db.to_csv('C:\OLIST\olist_orders_dataset_new.csv',index=False)
#orders_db.to_csv('/Users/tejaltandel/archive/olist_orders_dataset.csv',index=False)

In [62]:
order_items_db['order_id'].nunique()

98666

In [63]:
order_payments_db['order_id'].nunique()

99440

# LOADING ALL DATASETS

### The pre-processed datasets will now be loaded

In [86]:
# olist_orders = pd.read_csv('C:\OLIST\olist_orders_dataset_new.csv')
# olist_products = pd.read_csv('C:\OLIST\olist_products_dataset_new.csv')
# olist_items = pd.read_csv('C:\OLIST\olist_order_items_dataset.csv')
# olist_customers = pd.read_csv('C:\OLIST\olist_customers_dataset_new.csv')
# olist_payments = pd.read_csv('C:\OLIST\olist_order_payments_dataset.csv')
# olist_sellers = pd.read_csv('C:\OLIST\olist_sellers_dataset_new.csv')
# olist_geolocation = pd.read_csv('C:\OLIST\olist_geolocation_dataset.csv')
# olist_reviews = pd.read_csv('C:\OLIST\olist_order_reviews_dataset_new.csv')
# olist_product_category_name = pd.read_csv('C:\OLIST\product_category_name_translation.csv')


olist_customers = pd.read_csv("/Users/tejaltandel/archive/olist_customers_dataset.csv")
olist_geolocation = pd.read_csv("/Users/tejaltandel/archive/olist_geolocation_dataset.csv")
olist_items = pd.read_csv("/Users/tejaltandel/archive/olist_order_items_dataset.csv")
olist_payments = pd.read_csv("/Users/tejaltandel/archive/olist_order_payments_dataset.csv")
olist_reviews = pd.read_csv("/Users/tejaltandel/archive/olist_order_reviews_dataset.csv")
olist_orders = pd.read_csv("/Users/tejaltandel/archive/olist_orders_dataset.csv")
olist_products = pd.read_csv("/Users/tejaltandel/archive/olist_products_dataset.csv")
olist_sellers = pd.read_csv("/Users/tejaltandel/archive/olist_sellers_dataset.csv")
olist_product_category_name = pd.read_csv("/Users/tejaltandel/archive/product_category_name_translation.csv")

# MERGING DATASETS INTO SINGLE DATAFRAME

In [88]:
all_data = olist_orders.merge(olist_items, on='order_id', how='left')
all_data = all_data.merge(olist_payments, on='order_id', how='inner')
all_data = all_data.merge(olist_reviews, on='order_id', how='inner')
all_data = all_data.merge(olist_products, on='product_id', how='inner')
all_data = all_data.merge(olist_customers, on='customer_id', how='inner')
all_data = all_data.merge(olist_sellers, on='seller_id', how='inner')
all_data = all_data.merge(olist_product_category_name,on='product_category_name',how='inner')

In [114]:
all_data.to_csv('/Users/tejaltandel/archive/olist_all_data.csv',index=False)

In [89]:
round((all_data.isnull().sum()/ len(all_data)*100),2)

order_id2                         0.00
order_id                          0.00
customer_id                       0.00
order_status                      0.00
order_purchase_timestamp          0.00
order_approved_at                 0.01
order_delivered_carrier_date      1.03
order_delivered_customer_date     2.08
order_estimated_delivery_date     0.00
order_item_id                     0.00
product_id                        0.00
seller_id                         0.00
shipping_limit_date               0.00
price                             0.00
freight_value                     0.00
payment_sequential                0.00
payment_type                      0.00
payment_installments              0.00
payment_value                     0.00
review_id2                        0.00
review_id                         0.00
review_score                      0.00
review_comment_title             88.06
review_comment_message           57.70
review_creation_date              0.00
review_answer_timestamp  

# PROCESSING OF DATE COLUMNS

In [91]:
date_columns = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date',
             'order_estimated_delivery_date', 'shipping_limit_date', 'review_creation_date', 'review_answer_timestamp'] 
for col in date_columns:
    all_data[col] = pd.to_datetime(all_data[col], format='%Y-%m-%d %H:%M:%S')

In [92]:
print('Duplicate data: ',all_data.duplicated().sum())

Duplicate data:  0


### Thus, we can conclude the dataframe has no duplicates 

In [94]:
# Handling empty input in the order_approved_at column
missing_1 = all_data ['order_approved_at'] - all_data ['order_purchase_timestamp']
print (missing_1.describe ())
print ('=' * 50)
print ('Median from time of order until approved:', missing_1.median ())

count                       115595
mean     0 days 10:30:11.553440892
std      0 days 22:15:50.693403196
min                0 days 00:00:00
25%                0 days 00:12:57
50%                0 days 00:20:51
75%                0 days 14:58:51
max               60 days 10:51:59
dtype: object
Median from time of order until approved: 0 days 00:20:51


### Median is considered because certain products were approved immediately while few were approved after a max of 60 days

In [116]:
add_1 = all_data [all_data ['order_approved_at'].isnull()] ['order_purchase_timestamp'] + missing_1.median()
all_data ['order_approved_at'] = all_data ['order_approved_at'].fillna(add_1)

order_id2                             0
order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                     0
order_delivered_carrier_date       1195
order_delivered_customer_date      2400
order_estimated_delivery_date         0
order_item_id                         0
product_id                            0
seller_id                             0
shipping_limit_date                   0
price                                 0
freight_value                         0
payment_sequential                    0
payment_type                          0
payment_installments                  0
payment_value                         0
review_id2                            0
review_id                             0
review_score                          0
review_comment_title             101808
review_comment_message            66703
review_creation_date                  0


In [117]:
all_data[['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date']].head()

order_purchase_timestamp   order_approved_at order_delivered_carrier_date  \
0      2017-10-02 10:56:33 2017-10-02 11:07:15          2017-10-04 19:55:00   
1      2017-10-02 10:56:33 2017-10-02 11:07:15          2017-10-04 19:55:00   
2      2017-10-02 10:56:33 2017-10-02 11:07:15          2017-10-04 19:55:00   
3      2017-08-15 18:29:31 2017-08-15 20:05:16          2017-08-17 15:28:33   
4      2017-08-02 18:24:47 2017-08-02 18:43:15          2017-08-04 17:35:43   

  order_delivered_customer_date  
0           2017-10-10 21:25:13  
1           2017-10-10 21:25:13  
2           2017-10-10 21:25:13  
3           2017-08-18 14:44:43  
4           2017-08-07 18:30:01

# Handling empty input in the order_delivered_carrier_date column

In [118]:
missing_2 = all_data ['order_delivered_carrier_date'] - all_data ['order_approved_at']
print (missing_2.describe ())
print ('=' * 50)
print ('Median from the time apporved to submission:', missing_2.median ())

add_2 = all_data [all_data ['order_delivered_carrier_date']. isnull ()] ['order_approved_at'] + missing_2.median ()
all_data ['order_delivered_carrier_date'] = all_data ['order_delivered_carrier_date']. fillna(add_2)

count                       114414
mean     2 days 20:14:33.475291485
std      3 days 13:05:05.858433457
min            -172 days +18:44:38
25%                0 days 21:14:01
50%         1 days 20:14:35.500000
75%                3 days 15:19:37
max              107 days 01:16:38
dtype: object
Median from the time apporved to submission: 1 days 20:14:35.500000
order_id2                             0
order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                     0
order_delivered_carrier_date          0
order_delivered_customer_date      2400
order_estimated_delivery_date         0
order_item_id                         0
product_id                            0
seller_id                             0
shipping_limit_date                   0
price                                 0
freight_value                         0
payment_sequential                    

### Median is considered because some are sent within 21 hours of the time he was approved, some are up to 107 days

# Handling empty input in the order_delivered_customer_date column

In [121]:
missing_3 = all_data ['order_delivered_customer_date'] - all_data ['order_delivered_carrier_date']
print (missing_3.describe ())
print ('=' * 50)
print ('Median from the time it was sent to the customer received:', missing_3.median ())

add_3 = all_data [all_data ['order_delivered_customer_date']. isnull ()] ['order_delivered_carrier_date'] + missing_3.median ()
all_data ['order_delivered_customer_date'] = all_data ['order_delivered_customer_date']. fillna(add_3)


count                       115609
mean     9 days 02:59:51.941721665
std      8 days 11:17:48.302731257
min             -17 days +21:41:31
25%                4 days 02:32:25
50%                7 days 01:25:10
75%               11 days 19:02:02
max              205 days 04:35:00
dtype: object
Median from the time it was sent to the customer received: 7 days 01:25:10
order_id2                             0
order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                     0
order_delivered_carrier_date          0
order_delivered_customer_date         0
order_estimated_delivery_date         0
order_item_id                         0
product_id                            0
seller_id                             0
shipping_limit_date                   0
price                                 0
freight_value                         0
payment_sequential             

In [122]:
all_data

order_id2                          order_id  \
0        ON23382  e481f51cbdc54678b7cc49136f2d6af7   
1        ON23382  e481f51cbdc54678b7cc49136f2d6af7   
2        ON23382  e481f51cbdc54678b7cc49136f2d6af7   
3        ON86476  128e10d95713541c87cd1a2e48201934   
4        ON76683  0e7e841ddf8f8f2de2bad69267ecfbcf   
...          ...                               ...   
115604   ON90047  442a41b00a8a8bda35511fdf1eb65cc5   
115605   ON84440  1dbff949801376b8795a093d31cfe205   
115606   ON49582  6775b950cdc4da0ce39b261032f92c84   
115607   ON41795  73d60420cd1a179b2d8887d538efe4c2   
115608   ON14547  f5cf5716413185387030a378bdd46ebe   

                             customer_id order_status  \
0       9ef432eb6251297304e76186b10a928d    delivered   
1       9ef432eb6251297304e76186b10a928d    delivered   
2       9ef432eb6251297304e76186b10a928d    delivered   
3       a20e8105f23924cd00833fd87daa0831    delivered   
4       26c7ac168e1433912a51b924fbd34d34    delivered   
...                                  ...          ...   
115604  e42e2e8642f9c09100b6667e321c2811     invoiced   
115605  0cbd2df3d6e9d8b1eaf5cfefc44886bc    delivered   
115606  c4d135422be0d4333f506acde8af9419    delivered   
115607  65d9213dfa004c17dc126503e106e4a3    delivered   
115608  ff95df9a387c8d032a2ca6887cf77236   processing   

       order_purchase_timestamp   order_approved_at  \
0           2017-10-02 10:56:33 2017-10-02 11:07:15   
1           2017-10-02 10:56:33 2017-10-02 11:07:15   
2           2017-10-02 10:56:33 2017-10-02 11:07:15   
3           2017-08-15 18:29:31 2017-08-15 20:05:16   
4           2017-08-02 18:24:47 2017-08-02 18:43:15   
...                         ...                 ...   
115604      2018-07-11 22:27:32 2018-07-11 22:45:43   
115605      2018-05-16 14:50:54 2018-05-16 15:18:42   
115606      2018-06-13 12:19:26 2018-06-13 17:39:14   
115607      2018-08-07 10:52:58 2018-08-07 11:05:18   
115608      2017-02-19 22:31:47 2017-02-19 22:45:13   

       order_delivered_carrier_date order_delivered_customer_date  \
0           2017-10-04 19:55:00.000       2017-10-10 21:25:13.000   
1           2017-10-04 19:55:00.000       2017-10-10 21:25:13.000   
2           2017-10-04 19:55:00.000       2017-10-10 21:25:13.000   
3           2017-08-17 15:28:33.000       2017-08-18 14:44:43.000   
4           2017-08-04 17:35:43.000       2017-08-07 18:30:01.000   
...                             ...                           ...   
115604      2018-07-13 19:00:18.500       2018-07-20 20:25:28.500   
115605      2018-05-18 12:10:00.000       2018-05-21 17:14:08.000   
115606      2018-06-15 13:09:00.000       2018-06-16 17:48:30.000   
115607      2018-08-08 08:31:00.000       2018-08-22 17:02:23.000   
115608      2017-02-21 18:59:48.500       2017-02-28 20:24:58.500   

       order_estimated_delivery_date  order_item_id  ... productid_2  \
0                         2017-10-18            1.0  ...       25915   
1                         2017-10-18            1.0  ...       25915   
2                         2017-10-18            1.0  ...       25915   
3                         2017-08-28            1.0  ...       25915   
4                         2017-08-15            1.0  ...       25915   
...                              ...            ...  ...         ...   
115604                    2018-07-20            1.0  ...       30726   
115605                    2018-05-28            1.0  ...        8250   
115606                    2018-06-28            1.0  ...       25114   
115607                    2018-09-04            1.0  ...       17335   
115608                    2017-03-21            1.0  ...       22993   

                      customer_unique_id customer_zip_code_prefix  \
0       7c396fd4830fd04220f754e42b4e5bff                     3149   
1       7c396fd4830fd04220f754e42b4e5bff                     3149   
2       7c396fd4830fd04220f754e42b4e5bff                     3149   
3       3a51803cc0d012c3b5dc8b7528cb05f7  

# Cleaning up name columns, and engineering required columns

In [123]:

all_data['customer_city'] = all_data['customer_city'].str.title()
all_data['seller_city'] = all_data['seller_city'].str.title()
all_data['product_category_name'] = all_data['product_category_name'].str.title()
all_data['payment_type'] = all_data['payment_type'].str.replace('_', ' ').str.title()
all_data['product_category_name_english'] = all_data['product_category_name_english'].str.replace('_', ' ')
all_data['review_response_time'] = (all_data['review_answer_timestamp'] - all_data['review_creation_date']).dt.days
#Olist['delivery_against_estimated'] = (Olist['order_estimated_delivery_date'] - Olist['order_delivered_customer_date']).dt.days
#Olist['product_size_cm'] = Olist['product_length_cm'] * Olist['product_height_cm'] * Olist['product_width_cm']
all_data['order_purchase_year'] = all_data.order_purchase_timestamp.apply(lambda x: x.year)
all_data['order_purchase_month'] = all_data.order_purchase_timestamp.apply(lambda x: x.month)
all_data['order_purchase_dayofweek'] = all_data.order_purchase_timestamp.apply(lambda x: x.dayofweek)
all_data['order_purchase_hour'] = all_data.order_purchase_timestamp.apply(lambda x: x.hour)
all_data['order_purchase_day'] = all_data['order_purchase_dayofweek'].map({0:'Mon',1:'Tue',2:'Wed',3:'Thu',4:'Fri',5:'Sat',6:'Sun'})
all_data['order_purchase_mon'] = all_data.order_purchase_timestamp.apply(lambda x: x.month).map({1:'Jan',2:'Feb',3:'Mar',4:'Apr',5:'May',6:'Jun',7:'Jul',8:'Aug',9:'Sep',10:'Oct',11:'Nov',12:'Dec'})

In [124]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 115609 entries, 0 to 115608
Data columns (total 51 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id2                      115609 non-null  object        
 1   order_id                       115609 non-null  object        
 2   customer_id                    115609 non-null  object        
 3   order_status                   115609 non-null  object        
 4   order_purchase_timestamp       115609 non-null  datetime64[ns]
 5   order_approved_at              115609 non-null  datetime64[ns]
 6   order_delivered_carrier_date   115609 non-null  datetime64[ns]
 7   order_delivered_customer_date  115609 non-null  datetime64[ns]
 8   order_estimated_delivery_date  115609 non-null  datetime64[ns]
 9   order_item_id                  115609 non-null  float64       
 10  product_id                     115609 non-null  object        
 11  

In [125]:
#all_data.to_csv("C:\OLIST\OLIST_ALLDATA.csv")
all_data.to_csv("/Users/tejaltandel/archive/olist_all_data.csv")

In [134]:
Olist = pd.read_csv("/Users/tejaltandel/archive/olist_all_data.csv")

In [135]:
Olist = Olist.drop(columns=["order_approved_at", "order_delivered_customer_date", "order_delivered_carrier_date", "product_weight_g","product_name_lenght", 
                            "product_description_lenght", "product_photos_qty","product_length_cm", "product_height_cm", "product_width_cm",
                            "product_length_cm","review_comment_title", "review_comment_message"])

In [136]:
Olist = Olist.rename(columns={'product_category_name_english':'product_category_name', 'product_category_name':'product_category_name_old'})

In [137]:
Olist = Olist.drop(columns=['product_category_name_old'])

In [138]:
dfObj = pd.DataFrame(Olist['order_purchase_timestamp'])

In [131]:
import datetime
def datechange(dd):
    kk = datetime.datetime.strptime(dd, '%Y-%m-%d %H:%M:%S.%f')
    kk = pd.to_datetime(dd, infer_datetime_format=True)
    return kk

In [141]:
#Olist['order_purchase_timestamp'] = Olist['order_purchase_timestamp'].apply(datetime(dfObj))
Olist['order_purchase_timestamp'] = pd.to_datetime(Olist['order_purchase_timestamp'], infer_datetime_format=True)

In [142]:
#Olist['shipping_limit_date'] = Olist['shipping_limit_date'].apply(datechange)
Olist['shipping_limit_date'] = pd.to_datetime(Olist['shipping_limit_date'], infer_datetime_format=True)

Olist['review_creation_date'] = pd.to_datetime(Olist['review_creation_date'], infer_datetime_format=True)

Olist['review_answer_timestamp'] = pd.to_datetime(Olist['review_answer_timestamp'], infer_datetime_format=True)

Olist['order_estimated_delivery_date'] = pd.to_datetime(Olist['order_estimated_delivery_date'], infer_datetime_format=True)

# CONCLUSION

### Thus in data preprocessing the following were performed :
### >Dummy values were removed
### >Duplicates were checked and removed
### >Independent datasets were merged for further model training
### >Statistical analysis were performed on the individual datasets